src/fold/mod.rs

```diff
@@ -139,6 +140,23 @@ impl Fold for Lifetime {
     }
 }
 
+impl Fold for Substitution {
+    type Result = Substitution;
+    fn fold_with(&self, folder: &mut Folder, binders: usize) -> Result<Self::Result> {
+        let mut tys = BTreeMap::new();
+        let mut lifetimes = BTreeMap::new();
+
+        for (var, ty) in &self.tys {
+            tys.insert(*var, ty.fold_with(folder, binders)?);
```

Comment:
Right, that's exactly what I mean -- I find this iterator-chained version much harder to read...